### PR TITLE
update custom logger

### PR DESCRIPTION
**Pull request by smusali**
This `Pull Request` is for customizing and adding `timezone` information to our custom logger; i.e. instead of
```
> const utils = require('./lib/utils');
> utils.log('custom info log');
[191024 21:19:48] custom info log
> utils.log('custom error log');
[191024 21:19:50] custom error log
```
this `Pull Request` changes the behavior in this way:
```
> const utils = require('./lib/utils');
> utils.log('custom info log');
2019-10-24 21:22:29 -0400 [info] custom info log
> utils.log('custom error log', 'error');
2019-10-24 21:22:39 -0400 [error] custom error log
```

### CODE REVIEW

**Comment by weizou19** [comment]
something wrong reading `distro` file, unrelated to the changes

**Comment by johnpadilla** [comment]
@weizou19 That's a timestamp we don't support. Something you might want to be aware of in the future.

**Comment by weizou19** [comment]
Not sure if ingester supports time format of `2019-11-01T06:17:10-0700`. 
cc: @johnpadilla @jakedipity

**Comment by weizou19** [comment]
Not sure if ingester supports time format of `2019-11-01T06:17:10-0700`. 
cc: @johnpadilla @jakedipity

File: lib/utils.js [hunk 1]
```diff
@@ -56,8 +56,10 @@ const buildUrl = (host, port, ssl, endpoint) => {
 };
 
 // Custom Logger
-const log = (msg) => {
-    console.log('[' + new Date().toISOString().substring(2, 19).replace('T', ' ').replace(/[Z-]/g, '') + '] ' + msg);
+const log = (message, level) => {
+    let datetime = new Date();
+    datetime = `${datetime.toISOString().replace('T', ' ').split('.')[0]} ${datetime.toString().split(' ')[5].split('GMT')[1]}`;
+    console.log(`${datetime} [${(level || 'info')}] ${message}`);
```
Comment (weizou19) [inline lib/utils.js]: highly recommend just have unit test for this, use [mockdate](https://www.npmjs.com/package/mockdate) library to mock  timezone/date and validate results

Comment (smusali) [reply lib/utils.js]: We'll visit this in the next improvement - integration testing will cover this part - in the next improvement, if needed, whole unit testing stuff is gonna be visited

File: lib/utils.js [hunk 2]
```diff
@@ -56,8 +56,10 @@ const buildUrl = (host, port, ssl, endpoint) => {
 };
 
 // Custom Logger
-const log = (msg) => {
-    console.log('[' + new Date().toISOString().substring(2, 19).replace('T', ' ').replace(/[Z-]/g, '') + '] ' + msg);
+const log = (message, level) => {
+    let datetime = new Date();
+    datetime = `${datetime.toISOString().replace('T', ' ').split('.')[0]} ${datetime.toString().split(' ')[5].split('GMT')[1]}`;
```
Comment (jakedipity) [inline lib/utils.js]: This is not being built correctly. You're using the UTC parts to fill in the `YYYY-MM-DD HH:MM:SS` portion, but then using the local timezone offset which results in an invalid date.

Also this seems like a roundabout way of building the string, why not just access the individual portions e.g. `datetime.getDate()`, `datetime.getMonth()`, etc...

File: lib/windows-utilities.js [hunk 3]
```diff
@@ -13,9 +13,9 @@ const streamEventLog = (config) => {
         const winEvent = new WinEventLogProcessor({
             events: config.winevent
         });
-        log('Streaming logs from ' + config.winevent.join(', ') + ' events');
+        log('streaming logs from ' + config.winevent.join(', ') + ' events');
         winEvent.on('data', (logObjects) => {
-            debug('Number of log objects found: ' + logObjects.length);
+            debug('# of log objects found: ' + logObjects.length);
```
Comment (jakedipity) [inline lib/windows-utilities.js]: Don't appreciate this shorthand in a log statement - seems lazy and for what reason?

File: lib/k8s.js [hunk 4]
```diff
@@ -102,7 +102,7 @@ var processDataStream = function(data) {
             data = data.slice(data.indexOf('\r\n') + 2);
         }
 
-        debug('Frame length ' + length + ', chunk length ' + data.length);
+        debug('frame length: ' + length + '; chunk length: ' + data.length);
```
Comment (jakedipity) [inline lib/k8s.js]: Could use a template literal here

File: lib/k8s.js [hunk 5]
```diff
@@ -149,15 +149,15 @@ var processDataStream = function(data) {
 
 exports.init = function() {
     client = net.connect(config.SOCKET_PATH, function() {
-        log('Connected to ' + config.SOCKET_PATH);
+        log('connected to ' + config.SOCKET_PATH);
```
Comment (jakedipity) [inline lib/k8s.js]: Could use a template literal here

File: lib/k8s.js [hunk 6]
```diff
@@ -149,15 +149,15 @@ var processDataStream = function(data) {
...
         client.write('GET /containers/json ' + HTTP_HEADERS);
     });
-    client.on('error', function(err) {
-        log('Error connecting to ' + config.SOCKET_PATH + ': ' + err);
+    client.on('error', function(error) {
+        log('error connecting to ' + config.SOCKET_PATH + ': ' + error, 'error');
```
Comment (jakedipity) [inline lib/k8s.js]: Could use a template literal here

File: lib/k8s.js [hunk 7]
```diff
@@ -149,15 +149,15 @@ var processDataStream = function(data) {
...
     });
     client.on('data', processDataStream);
     client.on('end', function() {
-        log('Disconnected from ' + config.SOCKET_PATH);
+        log('disconnected from ' + config.SOCKET_PATH, 'warn');
```
Comment (jakedipity) [inline lib/k8s.js]: Could use a template literal here

File: lib/k8s.js [hunk 8]
```diff
@@ -67,7 +67,7 @@ var processAllContainers = function(containers) {
             storeLabels(container.Id, container.Labels);
         }
     });
-    log('Found ' + Object.keys(networkIdToLabels).length + ' pods, ' + numLabels + ' labels in ' + config.SOCKET_PATH);
+    log('found ' + Object.keys(networkIdToLabels).length + ' pods, ' + numLabels + ' labels in ' + config.SOCKET_PATH);
```
Comment (jakedipity) [inline lib/k8s.js]: Could use a template literal here

File: lib/file-utilities.js [hunk 9]
```diff
@@ -262,7 +263,7 @@ const streamAllLogs = (config, callback) => {
 
 // Gracefully Shutting Down
 const gracefulShutdown = (signal) => {
-    log('Got ' + signal + ' signal, shutting down...');
+    log('got ' + signal + ' signal, shutting down...', 'warn');
```
Comment (jakedipity) [inline lib/file-utilities.js]: Could use a template literal here

File: lib/file-utilities.js [hunk 10]
```diff
@@ -248,8 +249,8 @@ const streamAllLogs = (config, callback) => {
             }
         });
 
-        journalctl.stderr.on('data', (err) => {
-            log('Error reading from journalctl: ' + err.toString().trim());
+        journalctl.stderr.on('data', (error) => {
+            log('error reading from journalctl: ' + error.toString().trim(), 'error');
```
Comment (jakedipity) [inline lib/file-utilities.js]: Could use a template literal here

Comment (mikehu) [reply lib/file-utilities.js]: Actually, is there a reason for dropping everything to lower case for logging?

**Comment by mikehu** [comment]
is there a reason for all the logging to have lowercase now?

**Comment by smusali** [comment]
Thanks, @jakedipity, for all the comments!

No, @mikehu, there is no specific reason - just keep it consistent with `debug` log statements

File: lib/utils.js (hunk 2 above)
Comment (smusali) [reply lib/utils.js]: @jakedipity, updated! It's now correctly matching what `(new Date()).toString()` is showing in terms of local time and timezone

File: lib/windows-utilities.js (hunk 3 above)
Comment (smusali) [reply lib/windows-utilities.js]: Done!

File: lib/k8s.js (hunk 4 above)
Comment (smusali) [reply lib/k8s.js]: Done!

File: lib/k8s.js (hunk 5 above)
Comment (smusali) [reply lib/k8s.js]: Done!

File: lib/k8s.js (hunk 6 above)
Comment (smusali) [reply lib/k8s.js]: Done!

File: lib/k8s.js (hunk 7 above)
Comment (smusali) [reply lib/k8s.js]: Done!

File: lib/k8s.js (hunk 8 above)
Comment (smusali) [reply lib/k8s.js]: Done!

File: lib/file-utilities.js (hunk 9 above)
Comment (smusali) [reply lib/file-utilities.js]: Done!

File: lib/file-utilities.js (hunk 10 above)
Comment (smusali) [reply lib/file-utilities.js]: Done!

**Comment by smusali** [comment]
@mikehu @jakedipity @weizou19 can you guys do final round of feedback? Thanks in advance!

File: lib/utils.js [hunk 11]
```diff
@@ -56,8 +56,16 @@ const buildUrl = (host, port, ssl, endpoint) => {
...
-    console.log('[' + new Date().toISOString().substring(2, 19).replace('T', ' ').replace(/[Z-]/g, '') + '] ' + msg);
+const log = (message, level) => {
+    const dateObject = new Date();
+    const padStart = (number, expectedLength) => {
+        const strNumber = String(Math.abs(number)).padStart(expectedLength, '0');
+        return (number < 0) ? `-${strNumber}` : strNumber;
+    };
+    const date = `${dateObject.getFullYear()}-${padStart(dateObject.getMonth() + 1, 2)}-${padStart(dateObject.getDate(), 2)}`;
+    const time = `${padStart(dateObject.getHours(), 2)}:${padStart(dateObject.getMinutes(), 2)}:${padStart(dateObject.getSeconds(), 2)}`;
+    const timezone = padStart(dateObject.getTimezoneOffset() * -5 / 3, 4);
```
Comment (weizou19) [inline lib/utils.js]: Even though the calculation is correct, it's confusing to readers 🤣. How about 
```
dateObject.getTimezoneOffset() * -1  / 60 * 100
```

Comment (smusali) [reply lib/utils.js]: you are right! updated it to be `dateObject.getTimezoneOffset() * -100  / 60`

**Pull request merged**